File: Python/deque/deque.py
class DequeNode:
	def __init__(self, data):
		self.data = data
		self.next = None
		self.previous = None


class Deque:
	def __init__(self, elems = None):
		self._front = None
		self._back = None
		self._length = 0
		if elems:
			for e in elems:
				self.add_last(e)
	
	def __len__(self):
		return self._length
	
	def __contains__(self, e):
		return self.contains(e)
	
	def empty(self):
		return self._length == 0
	
	@property
	def first(self):
		return getattr(self._front, 'data', None)
	
	@property 
	def last(self):
		return getattr(self._back, 'data', None)
	
	@first.setter
	def first(self, e):
		if self._length > 0:
			self._front.data = e
		else:
			self.add_first(e)
	
	@last.setter
	def last(self, e):
		if self._length > 0:
			self._back.data = e
		else:
			self.add_last(e)
	
	def add_first(self, e):
		if self._length == 0:
			self._front = DequeNode(e)
			self._back = self._front
		else:
			temp = DequeNode(e)
			temp.next = self._front
			self._front.previous = temp
			self._front = temp
		self._length += 1
	
	def add_last(self, e):
		if self._length == 0:
			self._back = DequeNode(e)
			self._front = self._back
		else:
			temp = DequeNode(e)
			self._back.next = temp
			temp.previous = self._back
			self._back = temp
		self._length += 1
	
	def remove_first(self):
		data = self.first
		if self._front == self._back:
			self.clear()
		else:
			self._front = self._front.next
			self._front.previous = None
			self._length -= 1
		return data
	
	def remove_last(self):
		data = self.last
		if self._front == self._back:
			self.clear()
		else:
			self._back = self._back.previous
			self._back.next = None
			self._length -= 1
		return data
	
	def contains(self, e):
		current = self._front
		while current:
			if current.data == e:
				return True
			current = current.next
		return False
	
	def clear(self):
		self._front = None
		self._back = None
		self._length = 0

File: Ruby/deque/deque.rb
class DequeNode
   attr_accessor :data, :next
   
   def initialize(data)
		@data = data
		@next = nil
		@previous = nil
   end
end

class Deque
	attr_reader :size
	
	def initialize(elems = nil)
		@front = nil
		@back = nil
		@size = 0
		if elems
			elems.each { |e| add_last(e) }
		end
	end
	
	def length
		@size
	end
	
	def empty?
		@size == 0
	end
	
	def first
		@front&.data
	end
	
	def last
		@back&.data
	end
	
	def first=(e)
		if @size > 0
			@front.data = e
		else
			add_first(e)
		end
	end
	
	def last=(e)
		if @size > 0
			@back.data = e
		else
			add_last(e)
		end
	end
	
	def add_first(e)
		if @size == 0
			@front = DequeNode.new(e)
			@back = @front
		else
			temp = DequeNode.new(e)
			temp.next = @front
			@front.previous = temp
			@front = temp
		end
		@size += 1
	end
	
	def add_last(e)
		if @size == 0
			@back = DequeNode.new(e)
			@front = @back
		else
			temp = DequeNode.new(e)
			@back.next = temp
			temp.previous = @back
			@back = temp
		end
		@size += 1
	end
	
	def remove_first
		data = first
		if @front == @back
			clear()
		else
			@front = @front.next
			@front.previous = nil
			@size -= 1
		end
		data
	end
	
	def remove_last
		data = last
		if @front == @back
			clear()
		else
			@back = @back.previous
			@back.next = nil
			@size -= 1
		end
		data
	end
	
	def include?(e)
		current = @front
		while current
			return true if current.data == e
			current = current.next
		end
		false
	end
	
	def clear
		@front = nil
		@back = nil
		@size = 0
	end
end
